component/discovery/discovery.go

```diff
@@ -0,0 +1,67 @@
+package discovery
+
+import (
+	"context"
+	"time"
+
+	"github.com/grafana/agent/component/metrics/scrape"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+// RunDiscovery is a utility for consuming and forwarding target groups from a discoverer.
+// It will handle collating targets (and clearing), as well as time based throttling of updates.
+// The channel provided should be the same channel passed to the Discovery's Run method.
+func RunDiscovery(ctx context.Context, ch <-chan []*targetgroup.Group, f func([]scrape.Target)) {
+	cache := map[string]*targetgroup.Group{}
+
+	dirty := false
+
+	const maxChangeFreq = 5 * time.Second
+	// this should give us 2 seconds at startup to collect some changes before sending
+	var lastChange time.Time = time.Now().Add(-3 * time.Second)
+	for {
```

Comment:
This whole thing is pretty hard to follow for me. I'd like to simplify this and sacrifice the completeness of the throttling functionality for now just so it's easier for us to reason about.

Something like what Prometheus does:

```go
var (
 ticker      = time.NewTicker(5 * time.Second)
 queueUpdate = make(chan struct{}, 1)
)


for {
  select {
  case <-ctx.Done():
    return 
    
  case <-ticker.C:
    select {
    case <-queueUpdate: 
      // ... calculate new targets and send to f ...  
    default:
      // No targets queued; wait to next tick.
    }
    
  case groups := <-ch:
    // ... update cache ... 

    // Schedule an update for the next tick if the channel is non-empty.
    select {
    case queueUpdate <- struct{}{}:
    default:
      // Update already queued; no-op. 
    }  
  }
}
```